### PR TITLE
ETFE-2469: Add trader name/ERN and 'change trader link' to every page…

### PR DESCRIPTION
**Pull request by nickg-hmrc**
…, if user has multiple ERNs on their login

### CODE REVIEW

File: app/featureswitch/core/config/FeatureSwitchingModule.scala [hunk 1]
```diff
@@ -44,7 +44,12 @@ case object UserAllowList extends FeatureSwitch {
   override val displayName: String = "Enable the User Allow List"
 }
 
+case object StubGetTraderKnownFacts extends FeatureSwitch {
+  override val configName: String = "features.stub-get-trader-known-facts"
```
Comment (LewisMitchell) [inline app/featureswitch/core/config/FeatureSwitchingModule.scala:48]: This config will need to be set appropriately for the QA, Staging and Production environment

File: conf/application.conf [hunk 2]
```diff
@@ -45,36 +45,46 @@ play.http.errorHandler = "handlers.ErrorHandler"
...
+    contact-frontend {
+      protocol = http
+      host = localhost
+      port = 9250
+    }
+    emcs-tfe-reference-data {
+      protocol = http
+      host = localhost
+      port = 8312
+    }
```
Comment (LewisMitchell) [inline conf/application.conf:77]: emcs-tfe-reference-data is the same as reference-data further up, so either `emcs-tfe-reference-data` or `reference-data`  can be removed

File: test/connectors/referenceData/GetTraderKnownFactsConnectorSpec.scala [hunk 3]
```diff
@@ -0,0 +1,68 @@
...
+
+  lazy val app = applicationBuilder(userAnswers = None).build()
+
+  implicit lazy val hc: HeaderCarrier = HeaderCarrier()
+  implicit lazy val ec: ExecutionContext = app.injector.instanceOf[ExecutionContext]
+
+  lazy val appConfig = app.injector.instanceOf[AppConfig]
+  lazy val connector = new GetTraderKnownFactsConnector(mockHttpClient, appConfig)
+
+  "check" - {
```
Comment (LewisMitchell) [inline test/connectors/referenceData/GetTraderKnownFactsConnectorSpec.scala]:
```suggestion
  "getTraderKnownFacts" - {
```

File: test/views/components/ActiveTraderSpec.scala [hunk 4]
```diff
@@ -0,0 +1,83 @@
...
+        }
+
+        "link must exist" in {
+          val html = activeTrader(Some(traderInfo), appConfig)
+          val doc = Jsoup.parse(html.toString())
+
+          val link =  doc.select(linkSelector).first
+
+          (link.attr("href"), link.text) mustEqual ("http://localhost:8310/emcs-tfe", messagesForLanguage.changeTraderType)
+        }
```
Comment (LewisMitchell) [inline test/views/components/ActiveTraderSpec.scala]: Could these be collapsed into one test? At the moment, it has to generate 3 identical HTML components. If it were collapsed into one test you could do something like:
```suggestion
        "render the correct HTML" in {

          val html = activeTrader(Some(traderInfo), appConfig)
          val doc = Jsoup.parse(html.toString())
          val link =  doc.select(linkSelector).first

          doc.select(divSelector).size mustEqual 1
          doc.select(titleSelector).text mustEqual s"$traderName ($ern)"
          link.attr("href") mustEqual "http://localhost:8310/emcs-tfe"
          link.text mustEqual messagesForLanguage.changeTraderType
        }
```

**Review by LewisMitchell — CHANGES_REQUESTED** [review]
> * Config key `play.http.secret.key` can be removed from [test/resources/application.conf](https://github.com/hmrc/emcs-tfe-cancel-movement-frontend/blob/ETFE-2469/test/resources/application.conf#L17). This is already provided by Bootstrap

Could you address this as well please?

File: app/featureswitch/core/config/FeatureSwitchingModule.scala (hunk 1 above)
Comment (nickg-hmrc) [reply app/featureswitch/core/config/FeatureSwitchingModule.scala:48]: I've made some PRs here:
https://github.com/hmrc/app-config-base/pull/9706
https://github.com/hmrc/app-config-production/pull/20418
https://github.com/hmrc/app-config-qa/pull/20653
https://github.com/hmrc/app-config-staging/pull/14212

File: app/connectors/referenceData/GetTraderKnownFactsHttpParser.scala [hunk 5]
```diff
@@ -0,0 +1,60 @@
...
+
+  implicit class GetTraderKnownFactsReads(exciseRegistrationId: String) extends HttpReads[Either[ErrorResponse, Option[TraderKnownFacts]]] {
+    override def read(method: String, url: String, response: HttpResponse): Either[ErrorResponse, Option[TraderKnownFacts]] = {
+      response.status match {
+        case OK =>
+          response.validateJson match {
+            case Some(valid) => Right(Some(valid))
+            case None =>
+              logger.warn(s"[read] Bad JSON response from emcs-tfe-reference-data")
+              Left(JsonValidationError)
```
Comment (LewisMitchell) [inline app/connectors/referenceData/GetTraderKnownFactsHttpParser.scala:42]: Looks to be a missing test for this scenario

File: app/connectors/referenceData/GetTraderKnownFactsConnector.scala
Comment (LewisMitchell) [inline app/connectors/referenceData/GetTraderKnownFactsConnector.scala:1]: Could you add an integration test for this, please?

**Review by LewisMitchell — CHANGES_REQUESTED** [review]

**Review by LewisMitchell — APPROVED** [review]

**Pull request merged**